Issue 4022: Fix Controller gRPC list streams operation to return authorized streams

ravisharda opened the pull request:
**Change log description**  
Fix the issue that prevents Controller gRPC list streams operation from returning authorized streams.

**Purpose of the change**  
Fixes #4022.

**What the code does**  
1. Caches the current gRPC context during list streams operation and reuses it during filtering.
2. Adds a new test integration test class that verifies Controller list streams operation via clients. (`io.pravega.test.integration.ControllerGrpcListStreamsTest`). Several new tests were added here. 
3. Adds new custom auth handler, custom credentials and custom principal classes for testing purposes. 
4. Renames some auth-related classes, refactors some existing code, and adds a new class in order to simplify the auth handling code and make it easier to understand and reason about its workings. There were some changes required in that set of classes for the current issue and refactoring the code a bit made it easier to make those changes.
5. Add additional logging statements to make it easier to debug auth issues.

**How to verify it**  
* Run tests in `io.pravega.test.integration.ControllerGrpcListStreamsTest` class.
* All other tests should also pass.

shiveshr commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthContext.java:
```diff
@@ -0,0 +1,48 @@
+     */
+    public static AuthContext current() {
+        Principal principal = AuthInterceptor.AUTH_CONTEXT_TOKEN.get();
+        AuthInterceptor serverAuthInterceptor = AuthInterceptor.INTERCEPTOR_OBJECT.get();
```
AuthInterceptor.AUTH_CONTEXT_TOKEN is static.
If there are multiple concurrent grpc requests, wouldnt we get the same auth context token?

ravisharda replied:
It represents only the key used to index the `Principal` object stored in the current gRPC context, which remains the same, and so it is static. The value - the `Principal` object - is a different object for each gRPC call, so concurrent requests don't share the values.

I've renamed it and the second key as well, and have now described them through comments as shown in the following fragment. Hopefully, the comments reveal the intent better.

https://github.com/pravega/pravega/blob/ec4427f74dc71e46414628d68208561d5ad9b76b/controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthInterceptor.java#L38-L48

ravisharda commented:
> I have one clarification question on the approach. Some explanation about what auth token token is etc in the description would be helpful.

@shiveshr As suggested, I've now added some explanation. 
* Added some comments on the `AuthInterceptor` class to explain the keys used to index the objects stored in the current `io.grpc.Context` object. 
* Added some comments `AuthInterceptor.current(...)` method about the intent of the `Context.Key.get()` calls.
I hope that helps.

shiveshr commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthInterceptor.java:
```diff
@@ -21,25 +21,36 @@
+     * Represents the key used for indexing the {@link java.security.Principal} object stored in the current
+     * {@link io.grpc.Context}.
+     */
+    static final Context.Key<Principal> PRINCIPAL_OBJECT_KEY = Context.key(PRINCIPAL_KEY_NAME);
```
Ok i think i understand this now.
The comment says this is the `Key` for indexing the principal. 
So the key is static and constant. But from the context we can get the value against this key. 
So if i understand this correctly, we store principal and interceptor into the context against some known keys and then we can get the values from `AuthContext`.`key`.get().

ravisharda replied:
Yes, exactly! The gRPC Context API is a bit awkward in this respect. 

(I've explained the same here: https://github.com/pravega/pravega/pull/4037#discussion_r308529649)

shiveshr commented on controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java:
```diff
@@ -398,32 +404,61 @@ public void createScope(ScopeInfo request, StreamObserver<CreateScopeStatus> res
+
+        final AuthContext ctx;
+        if (this.grpcAuthHelper.isAuthEnabled()) {
+            ctx = AuthContext.current();
```
will AuthContext.current return the last authContext?
What if there were multiple interceptors. Will there be multiple authContexts?
If so, this AuthContext.current will not retrieve all the contexts, right?

ravisharda replied:
The logic here works in the presence of multiple auth interceptors. Let me explain this in detail. 

`AuthContext.current()` will return an object containing the `Principal` and `AuthHandler` objects. The `Principal` is directly retrieved from the current gRPC Context. The `AuthHandler` is retrieved from the `AuthInterceptor` instance present in the current gRPC Context. 

There may be multiple `AuthInterceptor` instances registered in the system. One `AuthInterceptor` is registered per `AuthHandler` implementation found in the classpath. This happens in `io.pravega.controller.server.rpc.auth.registerInterceptors()` when the gRPC Server is constructed.  

At runtime, for each gRPC call, the `Principal` and `AuthInterceptor` in turn are set by one of the multiple `AuthInterceptors`  that might be registered in the system. The one that will do it is the one that has an `Authhandler` that matches the auth method (`Bearer`, `Basic`, etc.) specified by the client. Please see the code below: 
https://github.com/pravega/pravega/blob/7a923ed9dac4688ff1eed13a11d876292f6f90c2/controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthInterceptor.java#L59

Hope this clarifies. 

Also, I had added a new test to verify a scenario involving multiple AuthHandler plugins. It is here: https://github.com/pravega/pravega/blob/7a923ed9dac4688ff1eed13a11d876292f6f90c2/test/integration/src/test/java/io/pravega/test/integration/ControllerGrpcListStreamsTest.java#L127

shiveshr commented on controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java:
```diff
@@ -398,32 +404,61 @@ public void createScope(ScopeInfo request, StreamObserver<CreateScopeStatus> res
-                delegationToken -> controllerService.listStreams(
-                        scopeName, request.getContinuationToken().getToken(), listStreamsInScopeLimit)
+                () -> {
+                        String result = this.grpcAuthHelper.checkAuthorization(
```
why cant we simply call checkAuthorization as it internally takes the same context as it is computed here --> AuthContext.current()

ravisharda replied:
The gRPC Context isn't available at this stage of the operation and therefore the `Principal` and `AuthInterceptor` (erstwhile `PravegaInterceptor`) objects cannot be retrieved from the gRPC Context at this point. Since those values were null, we were defaulting to assuming the user doesn't have access to the stream. This was the root cause for the list streams filtering not returning any streams in scope with auth enabled. 

To avoid the issue, we now cache the relevant `Principal` and `AuthHandler` objects in the `AuthContext` object during the commencement of the operation when the gRPC Context is available. We then use the cached `AuthContext` for filtering later in the same operation.

Hope that clarifies.

ravisharda commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthContext.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * Initiates a new instance using the current {@link io.grpc.Context}.
+     *u
```
Fixed.

ravisharda commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthContext.java:
```diff
@@ -0,0 +1,52 @@
+     */
+    public static AuthContext current() {
+        // Obtains the Principal object stored in the current gRPC Context
+        Principal principal = AuthInterceptor.PRINCIPAL_OBJECT_KEY.get();
```
The gRPC Context API is a bit awkward in this respect. While the key is static, when we invoke the `get()` on the key, the value is fetched from the current context in the gRPC scope, so the value changes for each gRPC request/scope.

https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.Key.html

I've also modified `AuthContext` to update the comments to explain this better.

ravisharda commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthHandlerManager.java:
```diff
@@ -136,7 +136,7 @@ public boolean authorize(String resource, Principal principal, String credential
      *
      * @param authHandler the {@code AuthHandler} implementation to register
-     * @Throws NullPointerException {@code authHandler} is null
+     * @throws NullPointerException {@code authHandler} is null
```
Personally, I am of the opinion (in agreement with your comment) that an `IllegalArgumentException` - and not `NullPointerException` - should be thrown when an argument is illegal (including the case where `null` is not allowed). `NullPointerException` on the other typically indicates that a variable inappropriately turned out to be a `null`, such as when a `null` is dereferenced. 

There are many who strongly advocate using a NullPointerException instead, for asserting that 'null' isn't allowed as an argument. Joshua Bloch in his famous book "Effective Java" states the following: 
> Arguably, every erroneous method invocation boils down to an illegal argument or state, but other exceptions are standardly used for certain kinds of illegal arguments and states. If a caller passes null in some parameter for which null values are prohibited, convention dictates that NullPointerException be thrown rather than IllegalArgumentException. Similarly, if a caller passes an out-of-range value in a parameter representing an index into a sequence, IndexOutOfBoundsException should be thrown rather than IllegalArgumentException.

A lot of code in our codebase throws NullPointerException to indicate that `null` is not allowed. For example, io.pravega.common.Exceptions.checkNotNullOrEmpty(), which is widely used, throws a NullPointerException if the argument is null. Similarly, `com.google.common.base.checkNotNull() throws a NullPointerException to reject a null input.

ravisharda commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthHandlerManager.java:
```diff
@@ -146,6 +146,7 @@ public synchronized void registerHandler(AuthHandler authHandler) {
     /**
      * Loads the custom implementations of the AuthHandler interface dynamically. Registers the interceptors with grpc.
      * Stores the implementation in a local map for routing the REST auth request.
+     *
```
Done.

ravisharda commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthContext.java:
```diff
@@ -0,0 +1,52 @@
+        Principal principal = AuthInterceptor.PRINCIPAL_OBJECT_KEY.get();
+
+        // Obtains the AuthInterceptor object stored in the current gRPC Context
+        AuthInterceptor serverAuthInterceptor = AuthInterceptor.AUTH_INTERCEPTOR_OBJECT_KEY.get();
```
https://github.com/pravega/pravega/pull/4037#discussion_r308529649 addresses this comment too.

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthContext.java:
```diff
@@ -0,0 +1,54 @@
+public class AuthContext {
+
+    @Getter
+    private Principal principal;
```
`final`

and below too.

ravisharda replied:
Done.

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/AuthContext.java:
```diff
@@ -0,0 +1,54 @@
+    @Getter
+    private AuthHandler authHandler;
+
+    private AuthContext(Principal principal, AuthHandler authHandler) {
```
`@NonNull` for both arguments or explain in your JavaDoc which arguments can be null.

ravisharda replied:
Nulls are allowed for both the params. I've added JavaDocs comments to express that, as suggested.

andreipaduroiu commented on controller/src/main/java/io/pravega/controller/server/rpc/auth/GrpcAuthHelper.java:
```diff
@@ -13,51 +13,66 @@
+                log.warn("AuthContext is null");
+            }
+            String message = String.format("Principal [%s] not allowed [%s] access for resource [%s]",
+                    ctx != null ? ctx.getPrincipal() : null, expectedLevel.toString(), resource);
```
no need to call `.toString` here

ravisharda replied:
Fixed.

andreipaduroiu commented on controller/src/test/java/io/pravega/controller/server/rpc/auth/ControllerGrpcAuthFocusedTest.java:
```diff
@@ -573,4 +614,50 @@ private static File createAuthFile() {
+    static class ResultObserver<T> implements StreamObserver<T> {
+        private T result = null;
+        private Throwable error;
+        private final AtomicBoolean completed = new AtomicBoolean(false);
```
You only update this inside a `synchronized` block. Either remove the `synchronized` or make this a simple boolean.

fpj replied:
What about making it a future and completing the future upon the call to `onCompleted`? This way we can block on the call to `get()` (Line 645), which would simplify the implementation of that method and of others as the `this.wait` call is no longer necessary.

ravisharda replied:
I liked the idea of using the future for the purpose and have modified the code to use it instead.  That change addresses both the comments. Thanks for the suggestion - the code looks much more simpler now!

andreipaduroiu commented on controller/src/test/java/io/pravega/controller/server/rpc/auth/ControllerGrpcAuthFocusedTest.java:
```diff
@@ -573,4 +614,50 @@ private static File createAuthFile() {
+            synchronized (this) {
+                while (!completed.get()) {
+                    try {
+                        this.wait();
```
It's bad practice to block while holding a lock.

Similarly above, you invoke `notifyAll` while holding the lock.

ravisharda replied:
The modified code doesn't use these.

andreipaduroiu commented on test/integration/src/main/resources/META-INF/services/io.pravega.auth.AuthHandler:
```diff
@@ -0,0 +1,10 @@
+#
```
Are these files needed?

ravisharda replied:
Yes, these files are necessary for the custom test auth handler to get loaded via the service loader.

fpj commented on test/integration/src/main/java/io/pravega/test/integration/auth/customplugin/CustomAuthHandler.java:
```diff
@@ -0,0 +1,62 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.test.integration.auth.customplugin;
```
If this is for test purposes, what about calling it `TestAuthHandler` rather than `CustomAuthHandler`?

ravisharda replied:
Sure, I've modified the class names accordingly.

fpj commented on test/integration/src/main/java/io/pravega/test/integration/auth/customplugin/CustomPrincipal.java:
```diff
@@ -0,0 +1,35 @@
+
+@Slf4j
+@ToString
+public class CustomPrincipal implements Principal {
```
Same comment about changing `Custom*`  to `Test*`.

ravisharda replied:
Done.

ravisharda commented on test/integration/src/main/java/io/pravega/test/integration/auth/customplugin/CustomAuthHandler.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (c) 2019 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Done.

fpj commented on controller/src/test/java/io/pravega/controller/server/rpc/auth/ControllerGrpcAuthFocusedTest.java:
```diff
@@ -627,37 +625,16 @@ public void onNext(T value) {
-                return result;
-            }
+        public T get() throws ExecutionException, InterruptedException {
+            return future.get();
```
Would it make sense to use `join` instead of `get` so that we don't have to declare that it throws?

ravisharda replied:
Yes, that's a good idea. Using `join()` instead of `get()` would make the code simpler here. I've modified it as suggested.
